util/src/timestamp.rs

```diff
@@ -1,22 +1,120 @@
-pub fn get_timestamp() -> i64 {
-    // Get UTC current datetime
-    let utc: DateTime<Utc> = Utc::now();
+/// NTP Timestamp information
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Clone)]
+pub struct TimestampNTP {
+    /// Last NTP timestamp
+    pub last_ntp: (i64, u32),
+    /// Difference between NTP and system timestamp
+    pub diff: Duration,
+    /// Flag to indicate if NTP is bigger or smaller
+    /// than system timestamp
+    pub bigger: bool,
+    /// Address to require NTP timestamp information
+    pub server: String,
+    /// Period to update global timestamp
+    pub update_period: i64,
+}
+
+lazy_static! {
+    static ref NTP_TS: RwLock<TimestampNTP> = RwLock::new(TimestampNTP {
+        server: String::from("0.pool.ntp.org:123"),
+        update_period: 600,
+        ..TimestampNTP::default()
+    });
+}
 
-    // Return number of non-leap seconds since Unix epoch
-    utc.timestamp()
+/// Get NTP timestamp
+pub fn get_global_timestamp() -> TimestampNTP {
+    NTP_TS.read().unwrap().clone()
+}
+
+/// Set NTP timestamp
+pub fn set_global_timestamp(ts: TimestampNTP) {
+    match NTP_TS.write() {
+        Ok(mut x) => {
+            if ts.last_ntp.0 < x.last_ntp.0 {
+                log::warn!(
+                    "Set new timestamp from the past. New is {}, old is {}",
+                    ts.last_ntp.0,
+                    x.last_ntp.0
+                );
+            } else {
+                log::debug!("Set new timestamp");
+            }
+            *x = ts;
+        }
+        Err(e) => {
+            log::error!("Failed to set timestamp: {}", e);
+        }
+    }
+}
+
+fn local_time(timestamp: ntp::protocol::TimestampFormat) -> chrono::DateTime<chrono::Local> {
+    let unix_time = ntp::unix_time::Instant::from(timestamp);
+    chrono::Local.timestamp(unix_time.secs(), unix_time.subsec_nanos() as _)
+}
+/// Get NTP timestamp from an addr specified
+pub fn get_timestamp_ntp(addr: &str) -> Option<(i64, u32)> {
+    ntp::request(addr)
+        .map(|p| {
+            let ts = local_time(p.receive_timestamp);
+
+            (ts.timestamp(), ts.timestamp_subsec_nanos())
+        })
+        .ok()
+}
+
+/// Function to get timestamp from system as UTC Unix timestamp, seconds since Unix epoch
+pub fn get_timestamp() -> (i64) {
+    get_timestamp_nanos().0
 }
 
 /// Function to get timestamp from system as UTC Unix timestamp, seconds and nanoseconds since Unix epoch
 pub fn get_timestamp_nanos() -> (i64, u32) {
     // Get UTC current datetime
     let utc: DateTime<Utc> = Utc::now();
 
-    // Return number of non-leap seconds since Unix epoch and the number of nanoseconds since the last second boundary
-    (utc.timestamp(), utc.timestamp_subsec_nanos())
+    let utc_secs = utc.timestamp();
+    let utc_subsec_nanos = utc.timestamp_subsec_nanos();
+
+    let mut old_ts = get_global_timestamp();
+
+    if utc_secs - old_ts.last_ntp.0 > old_ts.update_period {
```

Comment:
I would add a comment, something like

    // If the last request to a NTP server was more than update_period seconds ago,
    // query the server for a new timestamp

Reply:
And I think all the logic inside this "if" should be a separate function, for example `update_global_timestamp()`.